File: 99-Continue/157/preOrderTraversal.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.left = self.right = None
        self.data = data
        
class BinaryTree:
    def __init__(self):
        self.root=None
        
    def buildTree(self,arr:list[str])->Node:
        if len(arr)==0:
            return self.root
        queue = deque()
        self.root = Node(int(arr[0]))
        queue.append(self.root)
        i = 1
        while queue and i<len(arr):
            current = queue.popleft();
            if i<len(arr):
                if int(arr[i])!=-1:
                    current.left = Node(int(arr[i]))
                    queue.append(current.left)
                i+=1
            if i<len(arr):
                if int(arr[i])!=-1:
                    current.right = Node(int(arr[i]))
                    queue.append(current.right)
                i+=1
        return self.root
    
    def preOrderTraversal(self,root:Node):
        if root is not None:
            print(root.data,end=" ")
            self.preOrderTraversal(root.left)
            self.preOrderTraversal(root.right)
            
def main():
    tree = BinaryTree()
    strarr = input().split()
    tree.buildTree(strarr)
    tree.preOrderTraversal(tree.root)
    
if __name__=="__main__":
    main()
    

                
                
                
                
                
                
                
                
                
                
                
                

File: 99-Continue/157/PreOrderTraversal.java
import java.util.*;

public class PreOrderTraversal {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String str = sc.nextLine();
        
        String[] strarr = str.split(" ");
        
        BinaryTree tree = new BinaryTree();
        tree.buildTree(strarr);
        tree.preOrderTraversal(tree.root);
        sc.close();
    }
}


class Node{
    Node left,right;
    int data;
    Node(int val){
        this.data = val;
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root = null;
    }
    
    Node buildTree(String[] arr){
        if(arr.length==0){
            return null;
        }
        root = new Node(Integer.parseInt(arr[0]));
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        int i=1;
        while(!queue.isEmpty()&&i<arr.length){
            Node current = queue.poll();
            if(current.left==null){
                if(Integer.parseInt(arr[i])!=-1){
                    current.left = new Node(Integer.parseInt(arr[i]));
                    queue.offer(current.left);
                }
                i++;  
            }
            if(current.right==null){
                if(Integer.parseInt(arr[i])!=-1){
                    current.right = new Node(Integer.parseInt(arr[i]));
                    queue.offer(current.right);
                }
                i++;
            }
        }
        return root;
    }
    
    void preOrderTraversal(Node root){
        if(root!=null){
            System.out.print(root.data+" ");
            preOrderTraversal(root.left);
            preOrderTraversal(root.right);
        }
    }
    
}
